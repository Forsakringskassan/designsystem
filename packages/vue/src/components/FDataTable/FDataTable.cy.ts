import { defineComponent } from "vue";
import {
    DensityWrapper,
    densityWrapperHeight,
    densityWrapperWidth,
} from "@fkui/test-utils/vue";
import FTableColumn from "../FTableColumn/FTableColumn.vue";
import FDataTable from "./FDataTable.vue";

describe("density", () => {
    const DensityComponent = defineComponent({
        template: /* HTML */ `
            <density-wrapper>
                <f-data-table :rows="rows" striped key-attribute="id">
                    <template #caption> Tabell </template>
                    <template #default>
                        <f-table-column
                            v-for="column in columns"
                            :key="column.id"
                            title="Kolumnrubrik"
                            type="text"
                        >
                            Text
                        </f-table-column>
                    </template>
                </f-data-table>
            </density-wrapper>
        `,
        components: {
            DensityWrapper,
            FDataTable,
            FTableColumn,
        },
        data() {
            return {
                rows: ["1", "2", "3"].map((id) => ({ id })),
                columns: ["1", "2", "3"].map((id) => ({ id })),
            };
        },
    });

    /* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
    it.skip(`should be densified`, () => {
        cy.viewport(densityWrapperWidth, densityWrapperHeight);
        cy.mount(DensityComponent);
        cy.toMatchScreenshot();
    });
});
