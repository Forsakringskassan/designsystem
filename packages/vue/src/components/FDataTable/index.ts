export { default as FDataTable } from "./FDataTable.vue";
