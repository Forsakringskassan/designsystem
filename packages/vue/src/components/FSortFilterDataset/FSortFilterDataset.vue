<script setup lang="ts" generic="T">
import { type PropType, type Ref, computed, nextTick, onMounted, provide, ref, useTemplateRef, watch } from "vue";
import { focus, debounce, alertScreenReader, TranslationService } from "@fkui/logic";
import { IFlex, IFlexItem } from "../../internal-components/IFlex";
import { FSelectField } from "../FSelectField";
import { FIcon } from "../FIcon";
import { FTextField } from "../FTextField";
import { useTranslate } from "../../plugins";
import { FSortFilterDatasetSortCallback, FSortFilterDatasetMountCallback } from "./FSortFilterDatasetInterface";
import { sort } from "./FSortFilterSorter";
import { filter } from "./FSortFilterFilter";
import { type SortOrder } from "./sort-order";

const $t = useTranslate();

const searchString = ref("");
const defaultSortValue = { attribute: "", name: "", ascendingName: "", ascending: false, id: 0 };
const sortAttribute = ref<SortOrder>(defaultSortValue);
const sortFilterResult = ref<T[]>([]) as Ref<T[]>;
const debouncedFilterResultset = debounce(filterResultset, 250);

let tableCallbackOnSort: FSortFilterDatasetSortCallback = () => {
    /* do nothing */
};

let tableCallbackSortableColumns: FSortFilterDatasetMountCallback = () => {
    /* do nothing */
};

const props = defineProps({
    /**
     * The data that you wish to sort or filter.
     */
    data: {
        type: Array as PropType<T[]>,
        required: true,
        default: () => [],
    },
    /**
     * All the attributes you want to enable sorting for and the corresponding name to display in the dropdown.
     * Structured as `{attributeName: "Name for dropdown", secondAttributeName: "Name for dropdown"}`
     */
    sortableAttributes: {
        type: Object as PropType<Record<string, string>>,
        required: true,
    },
    /**
     * If set the data will be sorted by this attribute by default.
     */
    defaultSortAttribute: {
        type: String,
        required: false,
        default: () => "",
    },
    /**
     * Show/hides the sort dropdown.
     */
    showSort: {
        type: Boolean,
        required: false,
        default: true,
    },
    /**
     * Show/hides the filter input.
     */
    showFilter: {
        type: Boolean,
        required: false,
        default: true,
    },
    /**
     * Set placeholder text in filter input field.
     * Default is textkey "fkui.sort-filter-dataset.placeholder.filter"
     */
    placeholderFilter: {
        type: String,
        required: false,
        default: TranslationService.provider.translate("fkui.sort-filter-dataset.placeholder.filter", "Sök"),
    },
    /**
     * The order the data will be sorted by if defaultSortAttribute has been set.
     */
    defaultSortAscending: {
        type: Boolean,
        required: false,
        default: () => true,
    },
});

const emit = defineEmits<{
    /**
     * Emitted when the data is sorted.
     *
     * @arg items - The sorted data.
     */
    datasetSorted: [items: T[]];

    /**
     * Emits the used sorting attributes.
     *
     * @arg sortAttribute - The attributes used when sorting the data.
     */
    usedSortAttributes: [sortAttribute: SortOrder];
}>();

const showClearButton = computed(() => {
    return searchString.value.length > 0;
});

const sortOrders = computed((): SortOrder[] => {
    const arr = [] as SortOrder[];
    let id = 0;
    Object.keys(props.sortableAttributes).forEach((key: string) => {
        arr.push({
            attribute: key,
            name: props.sortableAttributes[key],
            ascendingName: $t("fkui.sort-filter-dataset.label.ascending", "stigande"),
            ascending: true,
            id: id++,
        });
        arr.push({
            attribute: key,
            name: props.sortableAttributes[key],
            ascendingName: $t("fkui.sort-filter-dataset.label.descending", "fallande"),
            ascending: false,
            id: id++,
        });
    });
    return arr;
});

const filterAttributes = computed(() => {
    return Object.keys(props.sortableAttributes);
});

provide("sort", (attribute: string, ascending: boolean) => {
    const foundAttribute = sortOrders.value.find((item) => {
        return item.attribute === attribute && item.ascending === ascending;
    });

    if (foundAttribute) {
        sortAttribute.value = foundAttribute;
    } else {
        sortAttribute.value = { attribute: "", ascending: false } as SortOrder;
    }

    sortFilterData();

    emit("usedSortAttributes", sortAttribute.value);
});

provide("registerCallbackOnSort", (callback: FSortFilterDatasetSortCallback) => {
    tableCallbackOnSort = callback;
});

provide("registerCallbackOnMount", (callback: FSortFilterDatasetMountCallback) => {
    tableCallbackSortableColumns = callback;
});

onMounted(() => {
    tableCallbackSortableColumns(Object.keys(props.sortableAttributes));
});

watch(
    () => props.data,
    () => {
        if (props.defaultSortAttribute !== "") {
            const foundAttribute = sortOrders.value.find((item) => {
                return item.attribute === props.defaultSortAttribute && item.ascending === props.defaultSortAscending;
            });
            if (foundAttribute) {
                sortAttribute.value = foundAttribute;
            }
        }
        sortFilterData();
    },
    { immediate: true, deep: true },
);

function sortFilterData(): void {
    const filteredData = filter(props.data, filterAttributes.value, searchString.value);

    if (sortAttribute.value.attribute === "") {
        sortFilterResult.value = filteredData;
    } else {
        sortFilterResult.value = sort([...filteredData], sortAttribute.value.attribute, sortAttribute.value.ascending);
    }
    // Await slot mount, otherwise if defaultSortAttribute is used the slot doesn't have time to register tableCallbackOnSort before this is called.
    nextTick(() => {
        tableCallbackOnSort(sortAttribute.value.attribute, sortAttribute.value.ascending);
    });
    emit("datasetSorted", sortFilterResult.value);
}

function onChangeSortAttribute(): void {
    sortFilterData();
    emit("usedSortAttributes", sortAttribute.value);
}

function onSearchInput(event: InputEvent): void {
    searchString.value = (event.target as HTMLInputElement).value;
    debouncedFilterResultset();
}

function onClickClearSearch(): void {
    searchString.value = "";
    sortFilterData();
    const input = useTemplateRef<HTMLElement>("search-field").value;
    focus(input);
}

function filterResultset(): void {
    sortFilterData();
    if (searchString.value === "") {
        alertScreenReader($t("fkui.sort-filter-dataset.aria-live.empty", "Sök redigera Sök tom"));
    } else {
        const searchAriaLive = $t(
            "fkui.sort-filter-dataset.aria-live.search",
            `Din sökning på "{{ search }}" gav {{ result }} träffar.`,
            {
                result: sortFilterResult.value.length,
                search: searchString.value,
            },
        );

        alertScreenReader(searchAriaLive);
    }
}
</script>

<template>
    <div class="sort-filter-dataset">
        <i-flex collapse gap="3x" wrap>
            <i-flex-item shrink align="center">
                <!--
                     @slot Slot for header to the left of the filter dropdown. Slot content is available through `v-slot="{ <propertyName> }"`, e.g. `v-slot="{ slotClass }"`.
                     @binding {string[]} slotClass - CSS classes to use for the SortFilter header element. Use with `:class="slotClass"`.
                -->
                <slot name="header" v-bind="{ slotClass: 'sort-filter-dataset__toolbar__header' }"></slot>
            </i-flex-item>

            <i-flex-item grow>
                <i-flex collapse float="right">
                    <i-flex-item v-if="showFilter" shrink align="center">
                        <div class="sort-filter-dataset__search">
                            <f-icon name="search" class="sort-filter-dataset__search__magnify-icon" />

                            <f-text-field
                                ref="search-field"
                                v-model="searchString"
                                inline
                                :placeholder="placeholderFilter"
                                maxlength="64"
                                @input="onSearchInput"
                            >
                                <span class="sr-only">{{ placeholderFilter }}</span>
                            </f-text-field>

                            <button
                                v-if="showClearButton"
                                type="button"
                                class="button button--discrete sort-filter-dataset__search__close-icon"
                                :title="$t('fkui.sort-filter-dataset.clear.filter', 'Rensa sökfält')"
                                @click="onClickClearSearch"
                            >
                                <f-icon name="close" />
                                <span class="sr-only">{{
                                    $t("fkui.sort-filter-dataset.clear.filter", "Rensa sökfält")
                                }}</span>
                            </button>
                        </div>
                    </i-flex-item>

                    <i-flex-item v-if="showSort" shrink align="center">
                        <f-select-field
                            v-model="sortAttribute"
                            class="sort-filter-dataset__sort"
                            inline
                            @change="onChangeSortAttribute"
                        >
                            <template #label>{{
                                $t("fkui.sort-filter-dataset.label.sort", "Sortera\u00A0på")
                            }}</template>

                            <option :value="defaultSortValue">
                                {{ $t("fkui.sort-filter-dataset.label.unsorted", "Välj") }}
                            </option>

                            <option v-for="sortOrder in sortOrders" :key="sortOrder.id" :value="sortOrder">
                                {{ sortOrder.name }} ({{ sortOrder.ascendingName }})
                            </option>
                        </f-select-field>
                    </i-flex-item>
                </i-flex>
            </i-flex-item>
        </i-flex>
        <!--
             @slot Slot for displaying the data. The data is available through `v-slot="{ <propertyName> }"`, e.g. `v-slot="{ sortFilterResult }"`.
             @binding {T[]} sortFilterResult - The sorted or filtered data.
        -->
        <slot v-bind="{ sortFilterResult }"></slot>
    </div>
</template>
