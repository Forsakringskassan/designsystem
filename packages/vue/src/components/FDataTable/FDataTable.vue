<script setup lang="ts" generic="T extends object">
import { type PropType, computed, onMounted, provide, ref } from "vue";
import { TableScroll, tableScrollClasses } from "../../utils";
import { getInternalKey, setInternalKeys } from "../../utils/internal-key";
import {
    FTableColumnData,
    FTableColumnSort,
    addColumn,
    setVisibilityColumn,
    updateSortOrder,
    setSortableColumns,
    getSortableIconName,
    getSortableIconClasses,
    FTableColumnType,
} from "../FTableColumn";
import { FSortFilterDatasetInjected } from "../FSortFilterDataset";
import { FIcon } from "../FIcon";
import { useTranslate } from "../../plugins";
import { useSlotUtils } from "../../composables";

const $t = useTranslate();
const { hasSlot } = useSlotUtils();
const { sort, registerCallbackOnSort, registerCallbackOnMount } = FSortFilterDatasetInjected();
const internalKey = getInternalKey<T>();

const columns = ref<FTableColumnData[]>([]);

defineOptions({
    inheritAttrs: false,
});

const props = defineProps({
    /**
     * The rows to be listed.
     * The rows will be listed in the given array order.
     */
    rows: {
        type: Array as PropType<T[]>,
        required: true,
    },
    /**
     * Unique attribute in rows.
     */
    keyAttribute: {
        type: String,
        required: false,
        default: undefined,
    },
    /**
     * If `true` alternating rows will use a different background color.
     */
    striped: {
        type: Boolean,
        default: false,
    },
    /**
     * Enable scrolling inside table.
     *
     * Can be one of the following values:
     *
     * - `"horizontal"`: Enables horizontal scrolling
     * - `"vertical"`: Enables vertical scrolling
     * - `"both"`: Enables scrolling in both directions
     * - `"none"`: Disables scrolling (default)
     */
    scroll: {
        type: String as PropType<TableScroll>,
        default: TableScroll.NONE,
        validator(value: string): boolean {
            const types: string[] = Object.values(TableScroll);
            return types.includes(value);
        },
    },
});

const hasCaption = computed(() => {
    return hasSlot("caption", {}, { stripClasses: [] });
});

const tableClasses = computed(() => {
    const classes = [];
    if (props.striped) {
        classes.push("table--striped");
    }
    return classes;
});

const isEmpty = computed(() => {
    return internalRows.value.length === 0;
});

const visibleColumns = computed((): FTableColumnData[] => {
    return columns.value.filter((col) => col.visible);
});

const wrapperClasses = computed(() => {
    return tableScrollClasses(props.scroll);
});

const tabindex = computed(() => {
    return props.scroll !== TableScroll.NONE ? 0 : undefined;
});

const internalRows = computed((): T[] => {
    const { keyAttribute } = props;
    if (keyAttribute) {
        return setInternalKeys(props.rows, keyAttribute as keyof T);
    }

    return setInternalKeys(props.rows);
});

provide("addColumn", (column: FTableColumnData) => {
    if (column.type === FTableColumnType.ACTION) {
        throw new Error("Cannot use action column in FDataTable component");
    }
    columns.value = addColumn(columns.value, column);
});

provide("setVisibilityColumn", (id: string, visible: boolean) => {
    setVisibilityColumn(columns.value, id, visible);
});

provide("textFieldTableMode", true);

provide(
    "renderColumns",
    computed(() => {
        return internalRows.value.length > 0;
    }),
);

onMounted(() => {
    registerCallbackOnSort(callbackOnSort);
    registerCallbackOnMount(callbackSortableColumns);
});

function rowKey(item: T): string {
    return String(item[internalKey]);
}

function columnClasses(column: FTableColumnData): string[] {
    const classes = ["table__column", `table__column--${column.type}`, column.size];
    if (column.sortable) {
        classes.push("table__column--sortable");
    }

    return classes;
}

function iconClasses(column: FTableColumnData): string[] {
    return getSortableIconClasses(column);
}

function iconName(column: FTableColumnData): string {
    return getSortableIconName(column);
}

function onClickColumnHeader(column: FTableColumnData): void {
    if (!column.sortable) {
        return;
    }

    let columnName = column.name;
    if (!columnName) {
        throw new Error("`FTableColumn` must have a unique `name` when used with `FSortFilterDataset`");
    }

    if (column.sort === FTableColumnSort.DESCENDING) {
        columnName = "";
        column.sort = FTableColumnSort.UNSORTED;
    }
    sort(columnName, column.sort !== FTableColumnSort.ASCENDING);
}

function callbackOnSort(columnName: string, ascending: boolean): void {
    updateSortOrder(columns.value, columnName, ascending);
}

function callbackSortableColumns(columnNames: string[]): void {
    setSortableColumns(columns.value, columnNames);
}

function escapeNewlines(value: string): string {
    return value.replace(/\n/g, "<br/>");
}
</script>

<template>
    <div :class="wrapperClasses">
        <table class="table" :class="tableClasses" :tabindex="tabindex" v-bind="$attrs">
            <caption v-if="hasCaption">
                <!-- @slot Slot for table caption. -->
                <slot name="caption"></slot>
            </caption>
            <colgroup>
                <col v-for="column in columns" :key="column.id" :class="column.size" />
            </colgroup>
            <thead>
                <tr class="table__row">
                    <slot v-bind="{ row: {} }" />

                    <th
                        v-for="column in visibleColumns"
                        :key="column.id"
                        scope="col"
                        :class="columnClasses(column)"
                        v-on="column.sortable ? { click: () => onClickColumnHeader(column) } : {}"
                    >
                        <!-- eslint-disable-next-line vue/no-v-html -->
                        <span v-html="escapeNewlines(column.title)"></span>
                        <f-icon v-if="column.sortable" :class="iconClasses(column)" :name="iconName(column)" />
                        <span v-if="column.description" class="table__column__description">{{
                            column.description
                        }}</span>
                    </th>
                </tr>
            </thead>
            <tbody>
                <tr v-if="isEmpty && columns.length === 0">
                    <slot v-bind="{ row: {} as T }"></slot>
                </tr>
                <tr v-if="isEmpty">
                    <td class="table__column table__column--action" :colspan="columns.length">
                        <!--
                             @slot Slot for displaying a message when table is empty. Default text is 'Tabellen är tom' (key fkui.data-table.empty).
                        -->
                        <slot name="empty">{{ $t("fkui.data-table.empty", "Tabellen är tom") }}</slot>
                    </td>
                </tr>
                <tr v-for="row in internalRows" :key="rowKey(row)" class="table__row">
                    <!--
                         @slot Slot for table row. The item object is available through `v-slot="{ <propertyName> }"`, e.g. `v-slot="{ row }"`.
                         @binding {T} row - The object to be visualized.
                    -->
                    <slot v-bind="{ row }" />
                </tr>
            </tbody>
        </table>
    </div>
</template>
