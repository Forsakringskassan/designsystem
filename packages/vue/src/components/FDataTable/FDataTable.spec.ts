import "html-validate/jest";
import { defineComponent } from "vue";
import { mount } from "@vue/test-utils";
import { FSortFilterDatasetMountCallback } from "../FSortFilterDataset";
import { FTableColumn } from "../FTableColumn";
import { TranslationPlugin } from "../../plugins";
import FDataTable from "./FDataTable.vue";

afterEach(() => {
    jest.restoreAllMocks();
});

describe("should match snapshot", () => {
    const TestComponent = {
        components: { FDataTable, FTableColumn },
        template: /* HTML */ `
            <f-data-table :rows="rows" key-attribute="a">
                <template #default="{ row }">
                    <f-table-column title="A" type="text">
                        {{ row.a }}
                    </f-table-column>
                    <f-table-column title="A" type="numeric">
                        {{ row.b }}
                    </f-table-column>
                </template>
            </f-data-table>
        `,
        data() {
            return {
                rows: [
                    { a: 1, b: 2 },
                    { a: 3, b: 4 },
                ],
            };
        },
        global: {
            plugins: [TranslationPlugin],
        },
    };

    it("thead", async () => {
        expect.assertions(1);
        const wrapper = mount(TestComponent);
        await wrapper.vm.$nextTick();
        expect(wrapper.find("thead")).toMatchSnapshot();
    });

    it("tbody", async () => {
        expect.assertions(1);
        const wrapper = mount(TestComponent);
        await wrapper.vm.$nextTick();
        expect(wrapper.find("tbody")).toMatchSnapshot();
    });

    it("should match snapshot with no items in rows and default text", async () => {
        const wrapper = mount({
            components: { FDataTable, FTableColumn },
            template: TestComponent.template,
            data() {
                return {
                    rows: [],
                };
            },
        });
        await wrapper.vm.$nextTick();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it("should match snapshot with no rows in table and custom text", async () => {
        const wrapper = mount({
            components: { FDataTable, FTableColumn },
            template: /* HTML */ `
                <f-data-table :rows="rows" key-attribute="a">
                    <template #default="{ row }">
                        <f-table-column title="A" type="text">
                            {{ row.a }}
                        </f-table-column>
                        <f-table-column title="A" type="numeric">
                            {{ row.b }}
                        </f-table-column>
                    </template>
                    <template #empty> THE LIST IS EMPTY! </template>
                </f-data-table>
            `,
            data() {
                return {
                    rows: [],
                };
            },
        });
        await wrapper.vm.$nextTick();
        expect(wrapper.html()).toMatchSnapshot();
    });
});

it("should add table colum headers to <thead> with correct classes", async () => {
    expect.assertions(6);
    const TestComponent = {
        components: { FDataTable, FTableColumn },
        template: /* HTML */ `
            <f-data-table :rows="rows" key-attribute="id">
                <template #default="{ row }">
                    <f-table-column title="A"></f-table-column>
                    <f-table-column title="B" shrink></f-table-column>
                </template>
            </f-data-table>
        `,
        data() {
            return {
                rows: [{ id: 1 }],
            };
        },
    };
    const wrapper = mount(TestComponent);
    await wrapper.vm.$nextTick();
    const tr = wrapper.findAll("thead tr");
    const th = tr[0].findAll("th");
    expect(tr).toHaveLength(1); /* should have a single <tr> element */
    expect(th).toHaveLength(2); /* should have one <th> per column */
    expect(th[0].text()).toBe("A");
    expect(th[1].text()).toBe("B");

    expect(th[0].classes().join(" ")).toBe(
        "table__column table__column--text table__column--expand",
    );

    expect(th[1].classes().join(" ")).toBe(
        "table__column table__column--text table__column--shrink",
    );
});

it("should throw error if `FTableColumn.name` is duplicated", async () => {
    expect.assertions(1);
    const TestComponent = {
        components: { FDataTable, FTableColumn },
        template: /* HTML */ `
            <f-data-table :rows="rows" key-attribute="id">
                <template #default="{ row }">
                    <f-table-column name="a" title="A"></f-table-column>
                    <f-table-column name="a" title="A"></f-table-column>
                </template>
            </f-data-table>
        `,
        data() {
            return {
                rows: [{ id: 1 }, { id: 2 }],
            };
        },
    };
    expect(() => {
        mount(TestComponent);
    }).toThrowErrorMatchingInlineSnapshot(
        `"Expected FTableColumn to have a unique name but encountered duplicate of "a""`,
    );
});

it("should set scope on table columns", async () => {
    expect.assertions(1);
    const TestComponent = {
        components: { FDataTable, FTableColumn },
        template: /* HTML */ `
            <f-data-table :rows="rows" key-attribute="id">
                <template #default="{ row }">
                    <f-table-column title="A"></f-table-column>
                </template>
            </f-data-table>
        `,
        data() {
            return {
                rows: [{ id: 1 }],
            };
        },
    };
    const wrapper = mount(TestComponent);
    await wrapper.vm.$nextTick();
    const th = wrapper.find("thead th");
    expect(th.attributes("scope")).toBe("col");
});

it("should throw exception when action is set on column", async () => {
    expect.assertions(1);
    const TestComponent = {
        components: { FDataTable, FTableColumn },
        template: /* HTML */ `
            <f-data-table :rows="rows" key-attribute="id">
                <template #default="{ row }">
                    <f-table-column title="A" type="action"></f-table-column>
                </template>
            </f-data-table>
        `,
        data() {
            return {
                rows: [{ id: 1 }],
            };
        },
    };
    const spy = jest.spyOn(console, "error").mockReturnValue(undefined);
    expect(() => mount(TestComponent)).toThrow(
        "Cannot use action column in FDataTable component",
    );
    spy.mockRestore();
});

it("should add <caption> if caption slot is set", async () => {
    expect.assertions(2);
    const TestComponent = {
        components: { FDataTable, FTableColumn },
        template: /* HTML */ `
            <f-data-table :rows="[]" key-attribute="id">
                <template #caption> My fancy caption </template>
            </f-data-table>
        `,
    };
    const wrapper = mount(TestComponent);
    await wrapper.vm.$nextTick();
    const caption = wrapper.find("caption");
    expect(caption.exists()).toBeTruthy();
    expect(caption.text()).toContain("My fancy caption");
});

it("should add <caption> if caption slot only contains sr-only text", async () => {
    expect.assertions(2);
    const TestComponent = {
        components: { FDataTable, FTableColumn },
        template: /* HTML */ `
            <f-data-table :rows="[]" key-attribute="id">
                <template #caption>
                    <span class="sr-only"> "My fancy caption </span>
                </template>
            </f-data-table>
        `,
    };
    const wrapper = mount(TestComponent);
    await wrapper.vm.$nextTick();
    const caption = wrapper.find("caption");
    expect(caption.exists()).toBeTruthy();
    expect(caption.text()).toContain("My fancy caption");
});

it("should not include caption if no caption slot", async () => {
    expect.assertions(1);
    const TestComponent = {
        components: { FDataTable, FTableColumn },
        template: /* HTML */ `
            <f-data-table :rows="[]" key-attribute="id"></f-data-table>
        `,
    };
    const wrapper = mount(TestComponent);
    await wrapper.vm.$nextTick();
    const caption = wrapper.find("caption");
    expect(caption.exists()).toBeFalsy();
});

it("should not add scroll classes when scroll property is unset", async () => {
    expect.assertions(1);
    const TestComponent = {
        components: { FDataTable, FTableColumn },
        template: /* HTML */ `
            <f-data-table :rows="[]" key-attribute="id"></f-data-table>
        `,
    };
    const wrapper = mount(TestComponent);
    await wrapper.vm.$nextTick();
    expect(wrapper.classes()).toEqual([]);
});

it("should set appropriate scroll class when scroll prop is given", async () => {
    expect.assertions(1);
    const TestComponent = {
        components: { FDataTable, FTableColumn },
        template: /* HTML */ `
            <f-data-table
                :rows="[]"
                key-attribute="id"
                scroll="horizontal"
            ></f-data-table>
        `,
    };
    const wrapper = mount(TestComponent);
    await wrapper.vm.$nextTick();
    expect(wrapper.classes()).toEqual([
        "table__scroll",
        "table__scroll--horizontal",
    ]);
});

it("should not add table--striped class unless striped is set", async () => {
    expect.assertions(1);
    const TestComponent = {
        components: { FDataTable, FTableColumn },
        template: /* HTML */ `
            <f-data-table :rows="[]" key-attribute="id"></f-data-table>
        `,
    };
    const wrapper = mount(TestComponent);
    await wrapper.vm.$nextTick();
    const table = wrapper.find("table");
    expect(table.classes()).not.toContain("table--striped");
});

it("should add table--striped class when striped is set", async () => {
    expect.assertions(1);
    const TestComponent = {
        components: { FDataTable, FTableColumn },
        template: /* HTML */ `
            <f-data-table
                :rows="[]"
                key-attribute="id"
                :striped="true"
            ></f-data-table>
        `,
    };
    const wrapper = mount(TestComponent);
    await wrapper.vm.$nextTick();
    const table = wrapper.find("table");
    expect(table.classes()).toContain("table--striped");
});

it("should be transparent", async () => {
    expect.assertions(1);
    const TestComponent = {
        components: { FDataTable },
        template: /* HTML */ `
            <f-data-table
                :rows="[]"
                key-attribute="id"
                foo="bar"
            ></f-data-table>
        `,
    };
    const wrapper = mount(TestComponent);
    await wrapper.vm.$nextTick();
    const table = wrapper.find("table");
    expect(table.attributes("foo")).toBe("bar");
});

it("should handle nested row objects no rows are present", async () => {
    expect.assertions(2);
    const TestComponent = defineComponent({
        components: { FDataTable, FTableColumn },
        template: /* HTML */ `
            <f-data-table :rows="[]" key-attribute="id">
                <template #caption> My fancy caption </template>
                <template #default="{ row }">
                    <f-table-column title="My Awesome Column">
                        {{ row.some.deeply.nested.prop }}
                    </f-table-column>
                </template>
            </f-data-table>
        `,
    });
    const wrapper = mount(TestComponent);
    await wrapper.vm.$nextTick();
    const th = wrapper.findAll("thead th");
    expect(th).toHaveLength(1);
    expect(th[0].text()).toBe("My Awesome Column");
});

it("should call provided sort method when clicking columnheader that is registrated as sortable", async () => {
    const mockedProvedSort = jest.fn();
    let registerSortableColumnsCallback: FSortFilterDatasetMountCallback;
    const TestComponent = {
        components: { FDataTable, FTableColumn },
        template: /*HTML*/ `
        <f-data-table :rows="[{ id: 1, name: 'some name' }]" key-attribute="id"  >
            <template #caption> My fancy caption </template>
            <template #default="{ row }">
                <f-table-column name="id" title="Id">
                    {{ row.id }}
                </f-table-column>
                <f-table-column name="name" title="Name">
                    {{ row.name }}
                </f-table-column>
            </template>
        </f-data-table>
    `,
        provide() {
            return {
                sort: mockedProvedSort,
                registerCallbackOnMount: (
                    callback: FSortFilterDatasetMountCallback,
                ) => {
                    registerSortableColumnsCallback = callback;
                },
            };
        },
    };
    const wrapper = mount(TestComponent);
    registerSortableColumnsCallback!(["name"]);
    await wrapper.vm.$nextTick();

    const colheaders = wrapper.findAll("thead tr th");
    const idColHeader = colheaders[0];
    const nameColHeader = colheaders[1];

    await idColHeader.trigger("click");
    expect(mockedProvedSort.mock.calls).toMatchInlineSnapshot(`[]`);

    await nameColHeader.trigger("click");
    expect(mockedProvedSort.mock.calls).toMatchInlineSnapshot(`
        [
          [
            "name",
            true,
          ],
        ]
    `);
});

describe("`keyAttribute`", () => {
    it("should not throw if valid and unique", async () => {
        expect.assertions(1);

        expect(() => {
            mount(FDataTable, {
                props: {
                    keyAttribute: "id",
                    rows: [{ id: "a" }, { id: "b" }, { id: "c" }],
                },
            });
        }).not.toThrow();
    });

    it("should throw error if not unique in items", async () => {
        expect.assertions(1);

        expect(() => {
            mount(FDataTable, {
                props: {
                    keyAttribute: "id",
                    rows: [{ id: "a" }, { id: "b" }, { id: "b" }],
                },
            });
        }).toThrowErrorMatchingInlineSnapshot(
            `"Expected each item to have key [id] with unique value but encountered duplicate of "b" in item index 2."`,
        );
    });

    it("should be optional", async () => {
        expect.assertions(1);

        expect(() => {
            mount(FDataTable, {
                props: {
                    rows: [{ id: "a" }, { id: "b" }, { id: "c" }],
                },
            });
        }).not.toThrow();
    });
});

describe("html-validate", () => {
    it("should require `key-attribute` to be non-empty if used", () => {
        expect.assertions(1);
        expect(
            '<f-data-table key-attribute=""></f-data-table>',
        ).not.toHTMLValidate({
            message: 'Attribute "key-attribute" has invalid value ""',
        });
    });

    it("should require row attribute", () => {
        expect.assertions(1);
        expect("<f-data-table></f-data-table>").not.toHTMLValidate({
            message: '<f-data-table> is missing required "rows" attribute',
        });
    });

    it("should require caption slot", () => {
        expect.assertions(1);
        expect("<f-data-table></f-data-table>").not.toHTMLValidate({
            message:
                '<f-data-table> component requires slot "caption" to be implemented',
        });
    });

    it("should require default slot", () => {
        expect.assertions(1);
        expect("<f-data-table></f-data-table>").not.toHTMLValidate({
            message:
                '<f-data-table> component requires slot "default" to be implemented',
        });
    });

    it.each`
        html
        ${'<f-email-text-field type="email" maxlength="80">E-post</f-email-text-field>'}
        ${'<f-email-text-field id="email-input" maxlength="80">E-post</f-email-text-field>'}
        ${'<f-email-text-field id="email-input" maxlength="80" v-validation.required>E-post</f-email-text-field>'}
    `("$html should be valid", ({ html }) => {
        expect.assertions(1);

        expect(html).toHTMLValidate();
    });
});
