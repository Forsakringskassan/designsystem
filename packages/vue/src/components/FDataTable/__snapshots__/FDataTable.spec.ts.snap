// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot should match snapshot with no items in rows and default text 1`] = `
<div class="">
  <table class="table">
    <!--v-if-->
    <colgroup>
      <col class="table__column--expand">
      <col class="table__column--expand">
    </colgroup>
    <thead>
      <tr class="table__row">
        <!--v-if-->
        <!--v-if-->
        <th scope="col" class="table__column table__column--text table__column--expand"><span>A</span>
          <!--v-if-->
          <!--v-if-->
        </th>
        <th scope="col" class="table__column table__column--numeric table__column--expand"><span>A</span>
          <!--v-if-->
          <!--v-if-->
        </th>
      </tr>
    </thead>
    <tbody>
      <!--v-if-->
      <tr>
        <td class="table__column table__column--action" colspan="2">Tabellen Ã¤r tom</td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`should match snapshot should match snapshot with no rows in table and custom text 1`] = `
<div class="">
  <table class="table">
    <!--v-if-->
    <colgroup>
      <col class="table__column--expand">
      <col class="table__column--expand">
    </colgroup>
    <thead>
      <tr class="table__row">
        <!--v-if-->
        <!--v-if-->
        <th scope="col" class="table__column table__column--text table__column--expand"><span>A</span>
          <!--v-if-->
          <!--v-if-->
        </th>
        <th scope="col" class="table__column table__column--numeric table__column--expand"><span>A</span>
          <!--v-if-->
          <!--v-if-->
        </th>
      </tr>
    </thead>
    <tbody>
      <!--v-if-->
      <tr>
        <td class="table__column table__column--action" colspan="2"> THE LIST IS EMPTY! </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`should match snapshot tbody 1`] = `
<tbody>
  <!--v-if-->
  <!--v-if-->
  <tr class="table__row">
    <td class="table__column table__column--text">1<span class="sr-only">&nbsp;</span></td>
    <td class="table__column table__column--numeric">2<span class="sr-only">&nbsp;</span></td>
  </tr>
  <tr class="table__row">
    <td class="table__column table__column--text">3<span class="sr-only">&nbsp;</span></td>
    <td class="table__column table__column--numeric">4<span class="sr-only">&nbsp;</span></td>
  </tr>
</tbody>
`;

exports[`should match snapshot thead 1`] = `
<thead>
  <tr class="table__row">
    <!--v-if-->
    <!--v-if-->
    <th scope="col" class="table__column table__column--text table__column--expand"><span>A</span>
      <!--v-if-->
      <!--v-if-->
    </th>
    <th scope="col" class="table__column table__column--numeric table__column--expand"><span>A</span>
      <!--v-if-->
      <!--v-if-->
    </th>
  </tr>
</thead>
`;
